pandajedi/jedicore/JediDBProxy.py

```diff
@@ -13692,3 +13692,222 @@ def extendSandboxLifetime_JEDI(self, jedi_taskid, file_name):
+    def queryTasksToBePending_JEDI(self, sql_query, params_map, reason):
+        comment = ' /* JediDBProxy.queryTasksToBePending_JEDI */'
+        methodName = self.getMethodName(comment)
+        methodName += " < sql={0} >".format(sql_query)
+        tmpLog = MsgWrapper(logger, methodName)
+        try:
+            self.conn.begin()
+            # sql to query
+            self.cur.execute(sql_query+comment, params_map)
+            taskIDs = self.cur.fetchall()
+            # sql to put the task in pending
+            sqlPDG = (  "UPDATE {0}.JEDI_Tasks "
+                        "SET lockedBy=NULL, lockedTime=NULL, "
+                            "status=:status, errorDialog=:err, "
+                            "modificationtime=CURRENT_DATE, oldStatus=status "
+                        "WHERE jediTaskID=:jediTaskID "
+                            "AND status IN ('ready','running','scouting') "
+                            "AND lockedBy IS NULL "
+                      ).format(jedi_config.db.schemaJEDI)
+            # loop over tasks
+            n_updated = 0
+            for jedi_taskid in taskIDs:
+                varMap = {}
+                varMap[':jediTaskID'] = jedi_taskid
+                varMap[':err'] = reason
+                varMap[':status'] = 'pending'
+                self.cur.execute(sqlPDG+comment, varMap)
+                nRow = self.cur.rowcount
+                if nRow == 1:
+                    self.record_task_status_change(jedi_taskid)
+                    n_updated += 1
+                elif nRow > 1:
+                    tmpLog.error('updated {0} rows with same jediTaskID'.format(nRow))
+            if not self._commit():
+                raise RuntimeError('Commit error')
+            tmpLog.debug('done with {0} rows'.format(n_updated))
+            # return
+            return n_updated
+        except Exception:
+            # roll back
+            self._rollback()
+            # error
+            self.dumpErrorMessage(tmpLog)
+            return None
+
+
+    # get  datasets of input and lib, to update data locality records
+    def get_tasks_inputdatasets_JEDI(self, vo):
+        comment = ' /* JediDBProxy.get_tasks_inputdatasets_JEDI */'
+        methodName = self.getMethodName(comment)
+        # last update time
+        methodName += " <vo={0}>".format(vo)
+        tmpLog = MsgWrapper(logger, methodName)
+        tmpLog.debug('start')
+        try:
+            retVal = None
+            # sql to get all jediTaskID and datasetID of input and lib
+            sql = ( "SELECT tabT.jediTaskID,datasetID, tabD.datasetName "
+                    "FROM {0}.JEDI_Tasks tabT,{0}.JEDI_Datasets tabD,{0}.JEDI_AUX_Status_MinTaskID tabA "
+                    "WHERE tabT.status=tabA.status AND tabT.jediTaskID>=tabA.min_jediTaskID AND tabT.jediTaskID=tabD.jediTaskID "
+                        "AND tabT.vo=:vo AND tabD.type IN ('input', 'lib')"
+                    ).format(jedi_config.db.schemaJEDI)
```

Comment:
Testing in production, the result of this query in notoriously large:
```
2020-11-24 15:11:51,224 panda.log.JediDBProxy: DEBUG    get_tasks_inputdatasets_JEDI <vo=atlas> start
2020-11-24 15:20:48,519 panda.log.JediDBProxy: DEBUG    get_tasks_inputdatasets_JEDI <vo=atlas> done with 42177839 rows
```
Sometimes it caused time-out
```
pandajedi.jedicore.Interaction.JEDITimeoutError: VO=any type=JEDITimeoutError stepIdx=4 : JediTaskBuffer.get_tasks_inputdatasets_JEDI did not get response for 600sec
```

Is there any way to shrink the result (say, can we omit data locality of tasks with final status ?) ?

Or, is there better way to separate tasks into batches?
Say, query every time with 10k-taskID range (taskid: 0~10k, 10k~20k, ...)?

Reply:
First of all, this is only for production tasks, so you can set task.prodsourcelabel, remove 'lib', and add "tabD.masterID is NULL", since basically we are interested in only primary input datasets for production tasks. Also, you may add a new timestamp column to jedi_tasks, update it every time you update data locality for the task, and add "tabT.newColumn is NULL or tabT.newColumn < :limit" in the query.